packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,331 @@
+      return 1 / 2;
+    case SpendType.SimpleSpend:
+      return 4 / 9;
+    case SpendType.UTXOFragmentation:
+      return 2 / 3 - 1 / numberOfOutputs;
+    case SpendType.Consolidation:
+      return 1 / numberOfInputs;
+    case SpendType.MixingOrCoinJoin:
+      let x = Math.pow(numberOfOutputs, 2) / numberOfInputs;
+      return ((2 / 3) * x) / (1 + x);
+    default:
+      throw new Error("Invalid spend type");
+  }
+}
+
+function determineSpendType(inputs: number, outputs: number): SpendType {
+  if (inputs === 1) {
+    if (outputs === 1) return SpendType.SweepSpend;
+    if (outputs === 2) return SpendType.SimpleSpend;
+    return SpendType.UTXOFragmentation;
+  } else {
+    if (outputs === 1) return SpendType.Consolidation;
+    return SpendType.MixingOrCoinJoin;
+  }
+}
+
+/*
+The transaction topology refers to the type of transaction based on 
+number of inputs and outputs.
+
+Expected Range : [0, 0.75]
+-> Very Poor : [0, 0.15]
+-> Poor : (0.15, 0.3]
+-> Moderate : (0.3, 0.45]
+-> Good : (0.45, 0.6]
+-> Very Good : (0.6, 0.75] 
+*/
+export async function privacyScoreByTxTopology(
+  transaction: Transaction,
+  client: BlockchainClient,
+): Promise<number> {
+  const numberOfInputs: number = transaction.vin.length;
+  const numberOfOutputs: number = transaction.vout.length;
+
+  const spendType: SpendType = determineSpendType(
+    numberOfInputs,
+    numberOfOutputs,
+  );
+  const score: number = spendTypeScores(
+    spendType,
+    numberOfInputs,
+    numberOfOutputs,
+  );
+
+  if (spendType === SpendType.Consolidation) {
+    return score;
+  }
+  for (let op of transaction.vout) {
+    let address = op.scriptPubkeyAddress;
+    let isResued = await isReusedAddress(address, client);
+    if (isResued === true) {
+      return score;
+    }
+  }
+  return score * DENIABILITY_FACTOR;
+}
+
+/* 
+In order to score for address reuse we can check the amount being hold by reused UTXOs 
+with respect to the total amount
+
+Expected Range : [0,1]
+-> Very Poor : (0.8, 1]
+-> Poor : [0.6, 0.8)
+-> Moderate : [0.4, 0.6)
+-> Good : [0.2, 0.4)
+-> Very Good : [0 ,0.2) 
+*/
+export async function addressReuseFactor(
+  utxos: AddressUtxos,
+  client: BlockchainClient,
+): Promise<number> {
+  let reusedAmount: number = 0;
+  let totalAmount: number = 0;
+
+  for (const address in utxos) {
+    const addressUtxos = utxos[address];
+    for (const utxo of addressUtxos) {
+      totalAmount += utxo.value;
+      let isReused = await isReusedAddress(address, client);
+      if (isReused) {
+        reusedAmount += utxo.value;
+      }
+    }
+  }
+  return reusedAmount / totalAmount;
+}
+
+async function isReusedAddress(
+  address: string,
+  client: BlockchainClient,
+): Promise<boolean> {
+  let txs: Transaction[] = await client.getAddressTransactions(address);
+  let countSend = 0;
+  let countReceive = 0;
+  for (const tx of txs) {
+    if(tx.isSend===true){
+      countSend++;
+    }else{
+      countReceive++;
+    }
```

Comment:
might as well just return true as soon as you hit either of these right?

Reply:
Well yes. Let's say with Send(S) and Receive(R) , we have following cases.
1. UTXO was received to address and is still not spent(S-0, R-1) - Address Not Reused
2. UTXO was received to address but it was spent already (S-1,  R-1) - Address Not Reused
3. UTXO was received more than once and (never spent, spent once, spent n times) - Address Reused
4. No UTXO was ever received to the address - Address Not resued

I think it makes more sense to just track the received count and if that is more than 1 we give true. I am changing the logic now.